libs/SalesforceSDK/src/com/salesforce/androidsdk/auth/NativeLoginManager.kt

```diff
@@ -266,6 +271,146 @@ internal class NativeLoginManager(
+        trimmedUsername.mapInvalidUsernameToResult()?.let { return StartRegistrationResult(it) }
+        val reCaptchaParameterGenerationResult = generateReCaptchaParameters(reCaptchaToken)
+
+        // Determine the OTP verification method.
+        val otpVerificationMethodString = otpVerificationMethod.identityApiAuthVerificationTypeHeaderValue
+        // Generate the start registration request body.
+        val startRegistrationRequestBodyString = runCatching {
+            StartRegistrationRequestBody(
+                recaptcha = reCaptchaParameterGenerationResult.nonEnterpriseReCaptchaToken,
+                recaptchaevent = reCaptchaParameterGenerationResult.enterpriseReCaptchaEvent,
+                userData = UserData(
+                    email = trimmedEmail,
+                    username = trimmedUsername,
+                    password = trimmedNewPassword,
+                    firstName = firstName.trim(),
+                    lastName = lastName.trim()
+                ),
+                otpVerificationMethod = otpVerificationMethodString
+            ).toJson()
+        }.onFailure { e ->
+            SalesforceSDKLogger.e(TAG, "Cannot JSON encode start registration request body due to an encoding error with message '${e.message}'.", e)
+            return StartRegistrationResult(UnknownError)
+        }.getOrNull()
+
+        // Create the start registration request.
+        val startRegistrationRequest = RestRequest(
+            POST,
+            LOGIN,
+            "$loginUrl$OAUTH_ENDPOINT_HEADLESS_INIT_REGISTRATION",
+            startRegistrationRequestBodyString,
+            null
+        )
+
+        // Submit the start registration request and fetch the response.
+        val startRegistrationResponse = suspendedRestCall(startRegistrationRequest) ?: return StartRegistrationResult(UnknownError)
+
+        // React to the start registration response.
+        return when (startRegistrationResponse.isSuccess) {
+            true -> {
+                startRegistrationResponse.consumeQuietly()
+                startRegistrationResponse.asJSONObject().let { response ->
+                    StartRegistrationResult(
+                        nativeLoginResult = Success,
+                        email = response.get("email").toString(),
+                        requestIdentifier = response.get("identifier").toString()
+                    )
+                }
+            }
+
+            else -> {
+                SalesforceSDKLogger.e(
+                    TAG,
+                    "Cannot start registration as the Salesforce Identity API returned an unsuccessful HTTP status. '${startRegistrationResponse.asString()}'"
+                )
+                StartRegistrationResult(UnknownError)
+            }
+        }
+    }
+
+    override suspend fun completeRegistration(
+        otp: String,
+        requestIdentifier: String,
+        otpVerificationMethod: OtpVerificationMethod
+    ) = submitAuthorizationRequest(
```

Comment:
Complete registration is identical to complete password-less authorization, so here's a common method extracted to fulfill both.